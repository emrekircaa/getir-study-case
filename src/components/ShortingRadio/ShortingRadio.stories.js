import { ShortingRadio } from "./index";

const meta = {
  title: "Components/ShortingRadio",
  component: ShortingRadio,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
};

export default meta;

export const Radio = {
  args: {
    selectedOptionId: 1,
    item: {
      id: 1,
    },
  },
};
