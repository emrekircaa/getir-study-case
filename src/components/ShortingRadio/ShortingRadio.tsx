import React, { useState, memo } from "react";
import styled from "styled-components";
import Tick from "../../assets/icons/Tick";
import { colors } from "../../theme";
import { useAppDispatch } from "../../hooks/useAppDistpatch";
import { useSelector } from "react-redux";
import { setFilter } from "../../store/filter/filterSlice";

type data = {
  id: string;
  label: string;
};
interface ShortingRadioProps {
  data?: data[];
  onSelectedRadioChange?: any;
  selectedOptionId?: boolean;
}
export const ShortingRadio: React.FC<ShortingRadioProps> = memo(
  ({ data, onSelectedRadioChange, selectedOptionId }) => {
    const [selectedRadioId, setSelectedRadioId] = useState(selectedOptionId);
    const dispatch = useAppDispatch();
    const filters = useSelector((store: any) => store.filter);

    const handleClick = (id: any) => {
      setSelectedRadioId(id);
      if (onSelectedRadioChange) onSelectedRadioChange(id);
      dispatch(setFilter({ ...filters, sort: id, page: 1 }));
    };
    return (
      <StyledRadiosContent>
        {data &&
          data.length > 0 &&
          data.map((item: any) => (
            <StyledRadioWrapper
              onClick={() => handleClick(item.id)}
              key={item.id}
              data-testid={`radio-${item.id}`}
            >
              <StyledRadio checked={item.id === selectedRadioId}>
                {item.id === selectedRadioId && <Tick />}
              </StyledRadio>
              <StyledRadioLabel>{item.label}</StyledRadioLabel>
            </StyledRadioWrapper>
          ))}
      </StyledRadiosContent>
    );
  }
);

const StyledRadiosContent = styled.div`
  display: flex;
  flex-direction: column;
  gap: 16px;
  margin-left: 24px;
  margin-right: 24px;
`;

const StyledRadioWrapper = styled.div`
  display: flex;
  align-items: center;
  cursor: pointer;
`;

const StyledRadio = styled.div<{ checked: boolean }>`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 22px;
  height: 22px;
  border-radius: 50%;
  background: white;
  border: 2px solid
    ${({ checked }) => (checked ? `${colors.primary}` : `${colors.dirtyWhite}`)};
  transition: 0.2s all;
`;

const StyledRadioLabel = styled.div`
  letter-spacing: 0.16px;
  font-size: 14px;
  line-height: 18px;
  margin-left: 12px;
  color: ${colors.gray500};
`;
