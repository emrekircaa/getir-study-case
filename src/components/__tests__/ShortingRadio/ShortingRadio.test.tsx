import { render, fireEvent } from "@testing-library/react";
import { ShortingRadio } from "../../ShortingRadio/ShortingRadio";

describe("ShortingRadio component", () => {
  const testData = [
    { id: 1, label: "Option 1" },
    { id: 2, label: "Option 2" },
    { id: 3, label: "Option 3" },
  ];

  it("renders with correct props", () => {
    const onSelectedRadioChange = jest.fn();
    const selectedOptionId = true;

    const { getByTestId } = render(
      <ShortingRadio
        data={testData}
        onSelectedRadioChange={onSelectedRadioChange}
        selectedOptionId={selectedOptionId}
      />
    );

    testData.forEach((item) => {
      const radioElement = getByTestId(`radio-${item.id}`);
      expect(radioElement).toMatchSnapshot();
    });
  });

  it("calls onSelectedRadioChange when a radio is clicked", () => {
    const onSelectedRadioChange = jest.fn();

    const { getByTestId } = render(
      <ShortingRadio
        data={testData}
        onSelectedRadioChange={onSelectedRadioChange}
      />
    );

    const radioToClick = testData[1];
    fireEvent.click(getByTestId(`radio-${radioToClick.id}`));

    expect(onSelectedRadioChange).toHaveBeenCalledTimes(1);
  });
});
