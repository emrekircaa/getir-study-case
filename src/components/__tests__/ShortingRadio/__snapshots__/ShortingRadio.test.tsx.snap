// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShortingRadio component renders with correct props 1`] = `
<div
  class="sc-gEvEer gtABzu"
  data-testid="radio-1"
>
  <div
    class="sc-eqUAAy fQFYTn"
  />
  <div
    class="sc-fqkvVR jzftmh"
  >
    Option 1
  </div>
</div>
`;

exports[`ShortingRadio component renders with correct props 2`] = `
<div
  class="sc-gEvEer gtABzu"
  data-testid="radio-2"
>
  <div
    class="sc-eqUAAy fQFYTn"
  />
  <div
    class="sc-fqkvVR jzftmh"
  >
    Option 2
  </div>
</div>
`;

exports[`ShortingRadio component renders with correct props 3`] = `
<div
  class="sc-gEvEer gtABzu"
  data-testid="radio-3"
>
  <div
    class="sc-eqUAAy fQFYTn"
  />
  <div
    class="sc-fqkvVR jzftmh"
  >
    Option 3
  </div>
</div>
`;
